### PR TITLE
Support ARM64 and simplify GitHub Action for building OCI images

### PR DESCRIPTION
**Pull request by Matthew-Beckett**
This PR adds QEMU and configures docker buildx to use an ARM64 emulation on AMD64 to build ARM64 targeted images on AMD64 runners.

### CODE REVIEW

File: .github/workflows/build-docker.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+name: Build Docker Images
+
+on: [push, pull_request]
+env:
+  TARGET_PLATFORMS: linux/amd64,linux/arm64
```
Comment (jodevsa) [inline .github/workflows/build-docker.yaml:5]: this isn't used anymore?

Comment (Matthew-Beckett) [reply .github/workflows/build-docker.yaml:5]: It is not feel free to remove

File: .github/workflows/build-docker.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+jobs: 
+  build-agent-docker:
+    strategy:
+      matrix:
+        platform:
+          - linux/amd64
```
Comment (jodevsa) [inline .github/workflows/build-docker.yaml:12]: What is the advantage of building an image per architecture versus one multi arch image?

File: .github/workflows/build-docker.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
+name: Build Docker Images
+
+on: [push, pull_request]
```
Comment (jodevsa) [inline .github/workflows/build-docker.yaml:3]: we are currently running the same job twice per PR. this PR ran 4 workflows (2 dulicates). Is that intended?

File: .github/workflows/build-docker.yaml (hunk 2 above)
Comment (Matthew-Beckett) [reply .github/workflows/build-docker.yaml:12]: The output artefact in ghcr.io gets updated with both architectures from both jobs it doesn't make multiple artefacts.

Also I noticed it was really slow to compile when doing parallel buildx building so I figured an action job per arch and image combo would be faster if only slightly.

File: .github/workflows/build-docker.yaml (hunk 3 above)
Comment (Matthew-Beckett) [reply .github/workflows/build-docker.yaml:3]: No, this should be on push to main and pull requests not just push.

**Comment by Matthew-Beckett** [comment]
Can I recommend due to the length of job we also configure a concurrency expression?

https://docs.github.com/en/actions/using-jobs/using-concurrency

**Comment by jodevsa** [comment]
> Can I recommend due to the length of job we also configure a concurrency expression?

Are you referring to lowering the number of parallel jobs running at the same time?

**Comment by jodevsa** [comment]
looks like the slowness is mainly coming from QEMU's emulation.

We can disable arm64 builds on PRs and only enable it for master just like how they did it here: https://github.com/DSpace/DSpace/pull/8315

**Comment by Matthew-Beckett** [comment]
> looks like the slowness is mainly coming from QEMU's emulation.
> 
> 
> 
> We can disable arm64 builds on PRs and only enable it for master just like how they did it here: https://github.com/DSpace/DSpace/pull/8315
> 
> 

I like this solution, happy to implement.

**Comment by Matthew-Beckett** [comment]
> > Can I recommend due to the length of job we also configure a concurrency expression?
> 
> 
> 
> Are you referring to lowering the number of parallel jobs running at the same time?

No, concurrency will group jobs together which are the same and can allow push to cancel previous in-process jobs in order to save runner minutes.

**Comment by jodevsa** [comment]
@Matthew-Beckett I did some tweaks like reordering the layers to achieve more caching. Somehow the builds now finish in 2-3 minutes for both architectures. Can you please review my changes and check if we are really building both platforms :D because it sounds too good to be true.

**Comment by Matthew-Beckett** [comment]
> @Matthew-Beckett I did some tweaks like reordering the layers to achieve more caching. Somehow the builds now finish in 2-3 minutes for both architectures. Can you please review my changes and check if we are really building both platforms :D because it sounds too good to be true.

It would appear it is in fact too good to be true, when checking the package in the GitHub UI you can see that the ARM and AMD64 commit hashes have diverged and are not on the same commit, also in the logs of the last run I don't see any linux/arm64 docker build only linux/amd64.

I don't think it's working properly.

https://github.com/jodevsa/wireguard-operator/pkgs/container/wireguard-operator%2Fmanager/89235169?tag=pr-38

**Comment by jodevsa** [comment]
I got it working!


I reworked the builds. latest run only took 6 minutes!


The speed bump is coming from not using emulation to build arm64 binaries. Currently i'm building the arm64 binaries over amd64 images. the changes are inspired from: https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/

**Comment by Matthew-Beckett** [comment]
LGTM!

**Comment by jodevsa** [comment]
fixes https://github.com/jodevsa/wireguard-operator/issues/20

**Pull request merged**